### ISSUE TITLE
OpenMoHAA for other computers/systems

### ISSUE THREAD

**Issue by gbudny** [openmoh/openmohaa#301]
Hi

It looks like someone started to work on OpenMoHAA for AmigaOS:

https://www.amigans.net/modules/newbb/viewtopic.php?topic_id=9488

I posted on the Pyra/OpenPandora forum, so you can probably look at some of the issues:

https://pyra-handheld.com/boards/threads/medal-of-honor-allied-assault-war-chest-on-pyra-openpandora.100011/#post-1718624

**Comment by smallmodel** [comment]
Interesting, I wanted to test OpenMoHAA on PowerPC as it's big-endian. I don't have the required hardware to test nor the time to set it up inside a QEMU VM.
It's interesting because the project was only tested on little-endian CPUs. There might be some place where endianness is not handled, which would cause crashes, the code must be fixed accordingly.

As for bison, the version of bison that is always used is 3.8.2, with flex 2.6.4.

**Comment by gbudny** [comment]
> Interesting, I wanted to test OpenMoHAA on PowerPC as it's big-endian. I don't have the required hardware to test nor the time to set it up inside a QEMU VM. It's interesting because the project was only tested on little-endian CPUs. There might be some place where endianness is not handled, which would cause crashes, the code must be fixed accordingly.

I didn't know it.

I don't have a computer with the PowerPC processor, but it's always great to see more software for it.

I could be wrong about it, but modern PowerPC computers are expensive: Amiga, Raptor, and IBM servers. It looks like that old Macs with PowerPC G5 are the most powerful and affordable computers.
  
On the other hand, I'm not sure if the OpenMoHHA built on G5 will work on modern Amiga computers or even less popular hardware like IBM or Raptor computers.

> As for bison, the version of bison that is always used is 3.8.2, with flex 2.6.4.

I posted your answer on the Pyra forum.

**Comment by gbudny** [comment]
Hi

Do you know what could cause this issue on Pyra?

> **Farox**
> Lowering the resolution (640*480) the game start...but something is wrong, terrain is not visible and if i walk into this...i fall down a bit under the objects of the scene.
> I try to change resolution but the game don't seems to see my changes...so i think to give up..

> **ptitSeb**
> It might be some "hardcoded" x86 behaviour... That might be difficult to track down. I remember have similar problems on Serious Sam...

You probably don't have this computer, but OpenMoHAA was ported to Linux ARM. In this case, maybe you can give Pyra users a hint of what they should check to fix this issue.

**Comment by samo790** [comment]
> > Interesting, I wanted to test OpenMoHAA on PowerPC as it's big-endian. I don't have the required hardware to test nor the time to set it up inside a QEMU VM. It's interesting because the project was only tested on little-endian CPUs. There might be some place where endianness is not handled, which would cause crashes, the code must be fixed accordingly.
> 
> I didn't know it.
> 
> I don't have a computer with the PowerPC processor, but it's always great to see more software for it.
> 
> I could be wrong about it, but modern PowerPC computers are expensive: Amiga, Raptor, and IBM servers. It looks like that old Macs with PowerPC G5 are the most powerful and affordable computers.    On the other hand, I'm not sure if the OpenMoHHA built on G5 will work on modern Amiga computers or even less popular hardware like IBM or Raptor computers.
> 
> > As for bison, the version of bison that is always used is 3.8.2, with flex 2.6.4.
> 
> I posted your answer on the Pyra forum.

Well an X5000 running on AmigaOS 4.1 have enough power to run Doom 3 :-)
https://m.youtube.com/watch?v=Nf6S86QiR3w&pp=ygUNRG9vbSAzIEFtaWdhIA%3D%3D

**Comment by samo790** [comment]
yes, prices of that hardware are quite high... but that's normal given that we're talking about a niche platform produced for enthusiasts ;-)
Anyway we were talking about hardware potential here, yes technically there would be no problems with openmohaa

**Comment by gbudny** [comment]
> Anyway we were talking about hardware potential here, yes technically there would be no problems with openmohaa

I never had doubts about it.

I hope that rjd324 will port OpenMoHAA to AmigaOS.

I remember that Halloween: The New Nightmare wasn't ported to AmigaOS/MorphOS because it used the BASS library. I remember that I wanted to see this game on these systems.

Unfortunately, no one decided to rewrite this audio part of the game to use SDL_mixer. That's an example of the issues that are probably more common for less popular systems like AmigaOS/MorphOS.

**Comment by samo790** [comment]
Yep this is a typical problem of niche platforms... we have various things, but some libraries will always be missing and not everything always works well the first time :-)

**Comment by smallmodel** [comment]
Some sections of the code were fixed to take endianness into account (for AmigaOS PowerPC). Only the server part was tested using a PPC64 big-endian machine on Debian 12. The client wasn't tested.
Pyra/OpenPandora uses an ARM32 CPU, but I have no idea whether or not big-endian is used.

> Lowering the resolution (640*480) the game start...but something is wrong, terrain is not visible and if i walk into this...i fall down a bit under the objects of the scene.

I had the same issue on PowerPC because terrain patches weren't read the way they should be in big-endian (values are stored as little-endian in the files)

**Comment by samo790** [comment]
@smallmodel 

Good job mate, this will semplify the work of the porting ;-)
Thank you

**Comment by gbudny** [comment]
@ptitSeb
> So, it's working, at least on the Pandora. There is still some work to do with the control, and the fullscreen, and the packaging, and the extraction of data from the GoG installer, and some crash at the exit that I should look into. So there yet. Speed seems correct, so that's a good news.

![image](https://github.com/openmoh/openmohaa/assets/13519111/25e48a92-2510-4079-a2f6-2be453eae7e1)

> I wont have time this weekend, will continue that next week...

https://pyra-handheld.com/boards/threads/medal-of-honor-allied-assault-war-chest-on-pyra-openpandora.100011/page-3#post-1718888

This is a great news!

**Comment by gbudny** [comment]
Hi

OpenMoHAA is available for AmigaOS and it was ported by rjd324:

https://www.amigans.net/modules/newbb/viewtopic.php?topic_id=9808

**Comment by smallmodel** [comment]
> Hi
> 
> OpenMoHAA is available for AmigaOS (PowerPC) and it was ported by rjd324:
> 
> https://www.amigans.net/modules/newbb/viewtopic.php?topic_id=9808
> 
> [@smallmodel](https://github.com/smallmodel) Do you plan to add the information about other systems like MorphOS or AmigaOS?
> 
> https://github.com/openmoh/openmohaa/blob/main/docs/getting_started_installation.md
> 
> I know some of these versions were ported by other users to the systems you don't use. However, OpenMoHAA has started to become a recognizable project, and it could help users find information about OpenMoHAA on their favorite systems. Maybe, more of these users will be interested to help testing and developing OpenMoHAA. I saw some suggestions and pull request from some of these users.
> 
> I'm thinking mainly about system requirements and the links to websites with these versions.

Those projects are both very valuable and interesting, it's great to see OpenMoHAA reaching more platforms.

That said, it looks like the source code of these ports (AmigaOS, Pyra and MorphOS) hasn't been made publicly available, despite openmohaa being licensed dunder the GPLv2 license. These ports may include custom changes deep into the source code, which could be integrated upstream and could benefit the wider community.
For reference, here's an example of open-source port (PSVita): https://github.com/Rinnegatamante/openmohaa

To help maintain transparency and collaboration, it would be ideal if forks were published, preferably on GitHub with a release section, so they can be easily referenced in the official documentation.

**Comment by gbudny** [comment]
Hi

Thank you for the answer.

> Those projects are both very valuable and interesting, it's great to see OpenMoHAA reaching more platforms.
> 
> That said, it looks like the source code of these ports (AmigaOS, Pyra and MorphOS) hasn't been made publicly available, despite openmohaa being licensed dunder the GPLv2 license. These ports may include custom changes deep into the source code, which could be integrated upstream and could benefit the wider community.

That's a slightly different issue.

I have to correct it. The functional port of OpenMoHAA for Pyra wasn't released. I know that @ptitSeb had some difficulties with it. I hope it will be released, but maybe he needs some help.
He is on GitHub so you can ask him.

I know that @Cowcat5150 planned to do it, but I don't know if a source code for MorphOS was released. I think he could answer it because he probably read it.

Lastly, I don't know @rjd324, but I saw that @samo790 posted a link to this issue on their forum. OpenMoHAA for AmigaOS isn't stable yet, so we should wait for a more playable version. They have a very active community on their forum.

> 
>For reference, here's an example of open-source port (PSVita): https://github.com/Rinnegatamante/openmohaa
 
I didn't know it, but that's cool!

> 
> To help maintain transparency and collaboration, it would be ideal if forks were published, preferably on GitHub with a release section, so they can be easily referenced in the official documentation.

I like this idea because it could help the OpenMoHAA project become more popular, and help solve many problems in the main branch. 

I hope that @Cowcat5150 and @rjd324 could share their opinion about it.

**Comment by Cowcat5150** [comment]
Soon the sources for MorphOS will be uploaded. 

The problem merging changes to the official branch is that partial code is located in its own "morphos" dir due to incompatibilities and/or missing functions: It's done that way because I didn't want to clutter the former code with dozens of "ifdefs" and workarounds needed for the MOS port (add here CMakeLists horror too). Needs its own stuff basically.

**Comment by 3246251196** [comment]
I can also upload. Will do soon.

**Comment by gbudny** [comment]
Hi

Great!

Thank you for the answers.

**Comment by smallmodel** [comment]
> Soon the sources for MorphOS will be uploaded.
> 
> The problem merging changes to the official branch is that partial code is located in its own "morphos" dir due to incompatibilities and/or missing functions: It's done that way because I didn't want to clutter the former code with dozens of "ifdefs" and workarounds needed for the MOS port (add here CMakeLists horror too). Needs its own stuff basically.

This is best practice to separate source files for OSes that deviate too much from generic implementation (unix and windows), `#ifdef` workarounds should definitely be avoided. If I understand correctly you created a separate `morphos` folder, I'd too create for example `sys/sys_morphos.c` and eventually add other source files like `sys/sys_morphos_some_stuff.c`.
OpenMoHAA also aims to target old hardware such as PowerPC-based hardware like Mac Mini G4 (which was an officially supported configuration), those changes could be integrated as long as they're different files, with some modifications to the CMake files, and as long as existing files aren't modified too much.

**Comment by Cowcat5150** [comment]
> If I understand correctly you created a separate morphos folder

Yes ....

> , I'd too create for example sys/sys_morphos.c and eventually add other source files like sys/sys_morphos_some_stuff.c.

That folder includes more than a couple of sys_morphos_whatever.c : From a necessary dll.a lib that doesn't exist in nowadays MOS SDK to a taylored crosscompile cmake script. Homebrew stuff that nobody wants in a multiplatform port.

Not so outrageous but as I said "it needs its own stuff" outside the project.

**Comment by 3246251196** [comment]
https://github.com/3246251196/openmohaa/tree/AmigaOS4_0_81_1

**Comment by samo790** [comment]
Could be merge that OS4 changes into the official repository?
